### PR TITLE
[FLINK-7357] [table] Created extended rules for WindowStartEndPropertiesRule on Having clause

### PR DESCRIPTION
**Pull request by walterddr**

### CODE REVIEW

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/sql/GroupWindowTest.scala [hunk 1]
```diff
@@ -143,4 +143,36 @@ class GroupWindowTest extends TableTestBase {
...
+            term("select", "rowtime, a")
+          ),
+          term("window", SlidingGroupWindow('w$, 'rowtime, 60000.millis, 900000.millis)),
+          term("select",
+            "COUNT(*) AS EXPR$0",
+            "SUM(a) AS $f1",
+            "start('w$) AS w$start",
+            "end('w$) AS w$end")
+        ),
+        term("select", "EXPR$0", "w$start")
```
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/sql/GroupWindowTest.scala]: Isn't your condition `> 0` missing in your logical plan?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/OverWindowITCase.scala [hunk 2]
```diff
@@ -836,6 +836,52 @@ class OverWindowITCase extends StreamingWithStateTestBase {
...
+    tEnv.registerTable("T1", t1)
+
+    val resultHopStartEndWithHaving = tEnv.sql(sqlQueryHopStartEndWithHaving).toAppendStream[Row]
+    resultHopStartEndWithHaving.addSink(new StreamITCase.StringSink[Row])
+
+    env.execute()
+
+    val expected = List(
+      "Hello,2,1970-01-01 03:53:00.0,1970-01-01 03:54:00.0",
+      "Hi,1,1970-01-01 03:53:00.0,1970-01-01 03:54:00.0"
```
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/OverWindowITCase.scala]: This result is not correct. If the count is `1` and we filter for sums over `b` larger than `1`, we should not include this result.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/FlinkRuleSets.scala [hunk 3]
```diff
@@ -147,7 +147,8 @@ object FlinkRuleSets {
 
     // Transform window to LogicalWindowAggregate
     DataSetLogicalWindowAggregateRule.INSTANCE,
-    WindowStartEndPropertiesRule.INSTANCE
+    WindowStartEndPropertiesRule.INSTANCE,
+    WindowStartEndPropertiesHavingRule.INSTANCE
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/FlinkRuleSets.scala]: Please also add tests for the batch logical plan.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/sql/GroupWindowTest.scala (hunk 1 above)
Comment (walterddr) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/sql/GroupWindowTest.scala]: Ahh. good catch. Bonehead, bonehead :-(. will fix it

**Comment by walterddr** [comment]
@twalthr looks like this should resolve the last filter issue. Please kindly take a look and see if this looks good to go in :-)

**Comment by fhueske** [comment]
Thanks for the PR @walterddr. 

We also need to replace group aux functions in the expressions of the filter predicate to support something like this:

```
HAVING 
  SUM(a) > 0 AND
  QUARTER(HOP_START(ts, INTERVAL '15' MINUTE, INTERVAL '1' MINUTE)) = 1
```

But that's just a minor change. 
I'll fix that, extend the tests to cover the case, and do some refactorings before I merge this PR.

Cheers, Fabian

**Pull request closed without merging**